tools/js_optimizer.py

```diff
@@ -81,127 +82,110 @@ def get_native_optimizer():
-            outs.append(out)
-            errs.append(err)
-          except OSError:
-            if compiler == shared.CLANG: raise # otherwise, OSError is likely due to g++ or clang++ not being in the path
-          if os.path.exists(output): return output
-        raise NativeOptimizerCreationException()
-
-      use_cmake_to_configure = WINDOWS # Currently only Windows uses CMake to drive the optimizer build, but set this to True to use on other platforms as well.
-      if use_cmake_to_configure:
-        return shared.Cache.get(name, create_optimizer_cmake, extension='exe')
-      else:
-        return shared.Cache.get(name, create_optimizer, extension='exe')
-    except NativeOptimizerCreationException as e:
-      shared.logging.debug('failed to build native optimizer')
-      handle_build_errors(outs, errs)
-      open(FAIL_MARKER, 'w').write(':(')
-      return None
-
-  def ignore_build_errors(outs, errs):
-    shared.logging.debug('to see compiler errors, build with EMCC_NATIVE_OPTIMIZER=g')
-  def show_build_errors(outs, errs):
-    for i in range(len(outs)):
-      shared.logging.debug('output from attempt ' + str(i) + ':\n' + shared.asstr(outs[i]) + '\n===========\n' + shared.asstr(errs[i]))
+            if WINDOWS and 'Visual Studio' in cmake_generator:
+              shutil.copyfile(os.path.join(build_path, cmake_build_type, 'optimizer.exe'), output)
+            else:
+              shutil.copyfile(os.path.join(build_path, 'optimizer'), output)
+            return output
+
+      raise NativeOptimizerCreationException()
+
+    def create_optimizer():
+      shared.logging.debug('building native optimizer: ' + name)
+      output = shared.Cache.get_path(name)
+      shared.try_delete(output)
+      for compiler in [shared.CLANG, 'g++', 'clang++']: # try our clang first, otherwise hope for a system compiler in the path
+        shared.logging.debug('  using ' + compiler)
+        try:
+          shared.run_process([compiler,
+                              shared.path_from_root('tools', 'optimizer', 'parser.cpp'),
+                              shared.path_from_root('tools', 'optimizer', 'simple_ast.cpp'),
+                              shared.path_from_root('tools', 'optimizer', 'optimizer.cpp'),
+                              shared.path_from_root('tools', 'optimizer', 'optimizer-shared.cpp'),
+                              shared.path_from_root('tools', 'optimizer', 'optimizer-main.cpp'),
+                              '-O3', '-std=c++11', '-fno-exceptions', '-fno-rtti', '-o', output] + args,
+                             stdout=log_output, stderr=log_output)
+        except Exception as e:
+          logging.debug(str(e))
+          continue # perhaps the later compilers will succeed
+        # success
+        return output
+      logging.critical('Failed to build native optimizer')
+      sys.exit(1)
```

Comment:
use `exit_with_error`

Reply:
Good idea, thanks - there's more places in the file that need to use that though, so let me do it as a followup PR which I'll write right now.